// @nxg-org/mineflayer-physics-util/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyToPlayerState = exports.convertPlayerState = exports.ControlStateHandler = exports.PlayerPoses = exports.PlayerState = exports.EntityState = exports.BotcraftPhysics = exports.EntityPhysics = exports.BaseSimulator = exports.PhysicsWorldSettings = exports.EPhysicsCtx = exports.PhysicsUtilWrapper = void 0;
exports.default = loader;
exports.initSetup = initSetup;
const settings_1 = require("./physics/settings");
const wrapper_1 = require("./wrapper");
function loader(bot) {
    if (!bot.physicsUtil) {
        initSetup(bot.registry);
        bot.physicsUtil = new wrapper_1.PhysicsUtilWrapper(bot);
    }
}
function initSetup(data) {
    settings_1.EPhysicsCtx.loadData(data);
}
var wrapper_2 = require("./wrapper");
Object.defineProperty(exports, "PhysicsUtilWrapper", { enumerable: true, get: function () { return wrapper_2.PhysicsUtilWrapper; } });
var settings_2 = require("./physics/settings");
Object.defineProperty(exports, "EPhysicsCtx", { enumerable: true, get: function () { return settings_2.EPhysicsCtx; } });
Object.defineProperty(exports, "PhysicsWorldSettings", { enumerable: true, get: function () { return settings_2.PhysicsWorldSettings; } });
var simulators_1 = require("./simulators");
Object.defineProperty(exports, "BaseSimulator", { enumerable: true, get: function () { return simulators_1.BaseSimulator; } });
var engines_1 = require("./physics/engines");
Object.defineProperty(exports, "EntityPhysics", { enumerable: true, get: function () { return engines_1.EntityPhysics; } });
Object.defineProperty(exports, "BotcraftPhysics", { enumerable: true, get: function () { return engines_1.BotcraftPhysics; } });
var states_1 = require("./physics/states");
Object.defineProperty(exports, "EntityState", { enumerable: true, get: function () { return states_1.EntityState; } });
Object.defineProperty(exports, "PlayerState", { enumerable: true, get: function () { return states_1.PlayerState; } });
Object.defineProperty(exports, "PlayerPoses", { enumerable: true, get: function () { return states_1.PlayerPoses; } });
var player_1 = require("./physics/player");
Object.defineProperty(exports, "ControlStateHandler", { enumerable: true, get: function () { return player_1.ControlStateHandler; } });
var physicsUtils_1 = require("./util/physicsUtils");
Object.defineProperty(exports, "convertPlayerState", { enumerable: true, get: function () { return physicsUtils_1.convertPlayerState; } });
Object.defineProperty(exports, "applyToPlayerState", { enumerable: true, get: function () { return physicsUtils_1.applyToPlayerState; } });
