// @nxg-org/mineflayer-physics-util/dist/wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhysicsUtilWrapper = void 0;
const engines_1 = require("./physics/engines");
const settings_1 = require("./physics/settings");
// /**
//  * Just a convenience thing.
//  */
// export enum SimulationTypes {
//     UNTIL_GROUND,
//     FOR_X_TICKS,
//     TO_DESTINATION,
// }
class PhysicsUtilWrapper {
    constructor(bot) {
        this.bot = bot;
        this.ePhysicsCtx = settings_1.EPhysicsCtx;
        this.data = bot.registry;
        settings_1.EPhysicsCtx.loadData(this.data);
        this.engine = new engines_1.EntityPhysics(this.data);
    }
    getPhysicsSim() {
        return new engines_1.EntityPhysics(this.data);
    }
    getPhysicsCtx(ctx, entity) {
        return settings_1.EPhysicsCtx.FROM_ENTITY(ctx, entity);
    }
    getPhysicsCtxRaw(ctx, entity, options = {}) {
        return settings_1.EPhysicsCtx.FROM_ENTITY_TYPE(ctx, entity, options);
    }
}
exports.PhysicsUtilWrapper = PhysicsUtilWrapper;
